packages/validators/src/url/utils.ts

```diff
@@ -1,8 +1,45 @@
+import { z } from 'zod'
+
 import { UrlValidationError } from '@/url/errors'
-import { isDynamicRoute } from '@/url/nextjs-dynamic-route'
-import { UrlValidatorWhitelist } from '@/url/options'
 
-const IS_NOT_HOSTNAME_REGEX = /[^.]+\.[^.]+/g
+/**
+ * Creates a Zod schema that validates a string to ensure all characters are within the allowed characters set.
+ *
+ * @param allowedChars - A string containing all allowed characters. If this is blank, the schema will always pass.
+ * @returns A Zod schema that validates a string to ensure all characters are within the allowed characters set.
+ *
+ * @example
+ * ```typescript
+ * const schema = createAllowedCharsSchema('abc123');
+ * schema.parse('abc'); // Valid
+ * schema.parse('a1b2c3'); // Valid
+ * schema.parse('abcd'); // Throws an error
+ * ```
+ */
+export const createAllowedCharsSchema = (allowedChars: string): z.ZodType<string> => {
+  if (!allowedChars) {
+    return z.string()
+  }
+  const allowed = new Set(Array.from(allowedChars))
+
+  const schema = z.string().refine(
+    (str: string) => {
+      for (const char of str) {
+        if (!allowed.has(char)) {
+          return false
+        }
+      }
+      return true
```

Comment:
```suggestion
      return Array.from(str).every(c=>allowed.has(c))
```

Would this more compact representation work as well? Only difference is creating an intermediate `Array`. Either way works though thanks to the `Set` optimisation.